racket-src/decl/decl-to-clause/trait-item.rkt

```diff
@@ -14,9 +14,82 @@
   ;; * The invariants that hold
   trait-item-decl-rules : CrateDecls CrateId (TraitId KindedVarIds_trait) TraitItem -> (Clauses Invariants)
 
-  [;; For a method declared in a trait: currently no logical rules are created.
-   (trait-item-decl-rules CrateDecls CrateId (TraitId KindedVarIds_trait) FnDecl)
-   ([] [])
+  [;; For a method declared in the trait Bar with parameter Self, like the following:
+   ;;
+   ;;     fn foo<T>(x: T) -> T where T: Ord { ... }
+   ;;
+   ;; We generate the following clauses
+   ;;
+   ;;     (∀ [(type Self) (type T)]
+   ;;         (well-formed-fn foo[Self T]) :-
+   ;;            (well-formed (type Self))
+   ;;            (well-formed (type T))
+   ;;            (is-implemented (Bar Self)))
```

Comment:
Do we need a `(well-formed (type foo[Self T])) => (is-implemented (Bar Self)))` invariant?

Reply:
We don't need this today, but I could see it becoming relevant in the future. Right now user's can't name these types explicitly, so they don't show up -- ah, well, hmm, maybe they can show up via inference! Something like this:

```rust
fn foo<T>() { }

fn bar() {
    let x = Some(foo).map(|f| ...);
}
```

inside that closure, `f` has a `FnDef` type. So in principle `(well-formed f)` would appear in its closure environment, and we could infer things based on that. Of course, anything we could infer should also be provable in the environment we inherit from `bar`. 

I am inclined to say NO invariants to start, but we should open an issue and compare against rustc I think.